text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const repository = ecr.Repository.fromRepositoryName(this, 'Repository', 'repo');
+const image = sagemaker.ContainerImage.fromEcrRepository(repository, 'tag');
+```
+
+#### `AssetImage`
+
+Reference a local directory containing a Dockerfile:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+import * as path from 'path';
+
+const image = sagemaker.ContainerImage.fromAsset(this, 'Image', {
+  directory: path.join('path', 'to', 'Dockerfile', 'directory')
+});
+```
+
+### Model Artifacts
+
+Models are often associated with model artifacts, which are specified via the `modelData` property
+which accepts a class that extends the `ModelData` abstract base class. The default is to have no
+model artifacts associated with a model.
+
+#### `S3ModelData`
+
+Reference an S3 bucket and object key as the artifacts for a model:
+
+```typescript
+import * as s3 from '@aws-cdk/aws-s3';
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const bucket = new s3.Bucket(this, 'MyBucket');
+const modelData = sagemaker.ModelData.fromBucket(bucket, 'path/to/artifact/file.tar.gz');
+```
+
+#### `AssetModelData`
+
+Reference local model data:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+import * as path from 'path';
+
+const modelData = sagemaker.ModelData.fromAsset(this, 'ModelData',
+  path.join('path', 'to', 'artifact', 'file.tar.gz'));
+```
+
+### `Model`
+
+The `Model` construct associates container images with their optional model data.
+
+#### Single Container Model
+
+In the event that a single container is sufficient for your inference use-case, you can define a
+single-container model:
+
+```typescript fixture=with-assets
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const model = new sagemaker.Model(this, 'PrimaryContainerModel', {
+  container: {
+    image: image,
+    modelData: modelData,
+  }
+});
+```
+
+#### Inference Pipeline Model
+
+An inference pipeline is an Amazon SageMaker model that is composed of a linear sequence of two to
+five containers that process requests for inferences on data. You use an inference pipeline to
+define and deploy any combination of pretrained Amazon SageMaker built-in algorithms and your own
+custom algorithms packaged in Docker containers. You can use an inference pipeline to combine
+preprocessing, predictions, and post-processing data science tasks. Inference pipelines are fully
+managed. To define an inference pipeline, you can provide additional containers for your model via
+the `extraContainers` property:
+
+```typescript fixture=with-assets
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const model = new sagemaker.Model(this, 'InferencePipelineModel', {
+  container: {
+    image: image1, modelData: modelData1
+  },
+  extraContainers: [
+    { image: image2, modelData: modelData2 },
+    { image: image3, modelData: modelData3 }
+  ],
```

Comment:
Oh I see.

Why the departure in syntax here, between `container` and `extraContainers` ? Why don't we have a single list of `containers`, which can have 1..* elements?